### PR TITLE
Display fluid localized name in IC2 fluid cell tooltip

### PR DESCRIPTION
**Pull request by MuXiu1997**
closed https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/10110

image

### CODE REVIEW

**Review by glowredman — COMMENTED** [review]
Why don't you modify the superclass directly so this applies to all subclasses which don't override it themselves?

**Comment by MuXiu1997** [comment]
I used @Redirect in superclass at first and it didn't work

**Comment by glowredman** [comment]
Try this:
```java
package com.mitchej123.hodgepodge.mixins.ic2;

import java.util.List;

import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Overwrite;

import ic2.core.item.ItemIC2FluidContainer;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.StatCollector;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.IFluidContainerItem;

@Mixin(ItemIC2FluidContainer.class)
public abstract class MixinItemIC2FluidContainer implements IFluidContainerItem {
	
	/**
	 * allows items to add custom lines of information to the mouseover description
	 * @author glowredman
	 * @author MuXiu1997
	 * @reason Use localized fluid name in the tooltip
	 */
	@SuppressWarnings({ "rawtypes", "unchecked" })
	@Overwrite
	public void addInformation(ItemStack itemStack, EntityPlayer player, List info, boolean b) {
		FluidStack fs = this.getFluid(itemStack);
		if(fs != null) {
			info.add("< " + fs.getLocalizedName() + ", " + fs.amount + " mB >");
		} else {
			info.add(StatCollector.translateToLocal("ic2.item.FluidContainer.Empty"));
		}
	}

}
```

**Comment by MuXiu1997** [comment]
I know, I didn't want to override it directly at first
I was just curious as to why the `@redirect` didn't work

**Comment by glowredman** [comment]
How did you try it?

**Comment by MuXiu1997** [comment]
I need to get up and turn on the computer.

**Comment by MuXiu1997** [comment]
```java
package com.mitchej123.hodgepodge.mixins.ic2;

import ic2.core.item.ItemIC2FluidContainer;
import net.minecraftforge.fluids.FluidStack;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Redirect;

@Mixin(ItemIC2FluidContainer.class)
public class MixinItemIC2FluidContainer {
    @Redirect(
            method = {
                    "addInformation(Lnet/minecraft/item/ItemStack;Lnet/minecraft/entity/player/EntityPlayer;Ljava/util/List;Z)V",
                    "getHudInfo(Lnet/minecraft/item/ItemStack;)Ljava/util/List;"
            },
            at = @At(
                    value = "INVOKE",
                    target = "Lnet/minecraftforge/fluids/FluidRegistry;getFluidName(Lnet/minecraftforge/fluids/FluidStack;)Ljava/lang/String;"
            ),
            remap = false
    )
    private String getFluidName(FluidStack fs) {
        return fs.getLocalizedName();
    }
}
```

**Comment by glowredman** [comment]
I believe it is because you didn't set `remap = false` in the `@At` annotation but in the `@Redirect` annotation. Thus, Mixins tries to remap `getFluidName` but not `addInformation` and `getHudInfo`.

**Comment by MuXiu1997** [comment]
```java
@Redirect(
            method = {
                    "addInformation(Lnet/minecraft/item/ItemStack;Lnet/minecraft/entity/player/EntityPlayer;Ljava/util/List;Z)V",
                    "getHudInfo(Lnet/minecraft/item/ItemStack;)Ljava/util/List;"
            },
            at = @At(
                    value = "INVOKE",
                    target = "Lnet/minecraftforge/fluids/FluidRegistry;getFluidName(Lnet/minecraftforge/fluids/FluidStack;)Ljava/lang/String;"
            )
    )
```
Compile failure

```java
@Redirect(
            method = {
                    "addInformation(Lnet/minecraft/item/ItemStack;Lnet/minecraft/entity/player/EntityPlayer;Ljava/util/List;Z)V",
                    "getHudInfo(Lnet/minecraft/item/ItemStack;)Ljava/util/List;"
            },
            at = @At(
                    value = "INVOKE",
                    target = "Lnet/minecraftforge/fluids/FluidRegistry;getFluidName(Lnet/minecraftforge/fluids/FluidStack;)Ljava/lang/String;",
                    remap = false
            ),
            remap = false

    )
```
Not work

**Comment by MuXiu1997** [comment]
```java
    @Redirect(
            method = "addInformation(Lnet/minecraft/item/ItemStack;Lnet/minecraft/entity/player/EntityPlayer;Ljava/util/List;Z)V",
            at = @At(
                    value = "INVOKE",
                    target = "Lnet/minecraftforge/fluids/FluidRegistry;getFluidName(Lnet/minecraftforge/fluids/FluidStack;)Ljava/lang/String;"
            )
    )
```
work

File: src/main/java/com/mitchej123/hodgepodge/mixins/ic2/MixinItemIC2FluidContainer.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Redirect;
+
+@Mixin(ItemIC2FluidContainer.class)
+public class MixinItemIC2FluidContainer {
+    @Redirect(
+            method = "addInformation(Lnet/minecraft/item/ItemStack;Lnet/minecraft/entity/player/EntityPlayer;Ljava/util/List;Z)V",
+            at = @At(
+                    value = "INVOKE",
+                    target = "Lnet/minecraftforge/fluids/FluidRegistry;getFluidName(Lnet/minecraftforge/fluids/FluidStack;)Ljava/lang/String;"
```
Comment (glowredman) [inline src/main/java/com/mitchej123/hodgepodge/mixins/ic2/MixinItemIC2FluidContainer.java:15]: This needs `remap = false` as `getFluidName()` is in a Forge class and thus not obfuscated

Comment (MuXiu1997) [reply src/main/java/com/mitchej123/hodgepodge/mixins/ic2/MixinItemIC2FluidContainer.java]: ok, modified

**Comment by MuXiu1997** [comment]
Thanks, added co-author

**Review by glowredman — APPROVED** [review]

**Pull request merged**